### ISSUE TITLE
[rllib] Misleading episodes_total metrics with Ape-X due to reporting only from selected workers

### ISSUE THREAD

**Issue by ElektroChan89** [ray-project/ray#9126]
### What is the problem?
It seems to be that when using Apex DQN in combination with the config parameter "use_exec_api=True", Apex DQN only gathers the batches of 1 worker instead of all workers.
In the example code below I use the Simple Corridor Environment with a print function in order to get the number of episodes of a single actor. 
Console output (results['episodes_total']) when "use_exec_api=True" indicates that Apex DQN takes the sample batches from one worker alone. 

*Ray version and other system information (Python version, TensorFlow version, OS):*
ray 0.8.5
python 3.6.9
Ubuntu 18.04.4

### Reproduction (REQUIRED)
```python
import numpy as np
import gym
from gym.spaces import Discrete, Box

import ray
from ray.rllib import agents


class SimpleCorridor(gym.Env):
    """Example of a custom env in which you have to walk down a corridor.
    You can configure the length of the corridor via the env config."""

    def __init__(self, config):
        self.end_pos = config["corridor_length"]
        self.cur_pos = 0
        self.action_space = Discrete(2)
        self.observation_space = Box(0.0, 999.0, shape=(1, ), dtype=np.float32)
        self.episodes = 0
        self.dont_count_begin_flag = True

    def set_corridor_length(self, length):
        self.end_pos = length
        print("Updated corridor length to {}".format(length))

    def reset(self):
        if not self.dont_count_begin_flag:
            self.episodes += 1
            print(f"Completed {self.episodes} episodes.")
        self.dont_count_begin_flag = False
        self.cur_pos = 0.0
        return [self.cur_pos]

    def step(self, action):
        assert action in [0, 1], action
        if action == 0 and self.cur_pos > 0:
            self.cur_pos -= 1.0
        elif action == 1:
            self.cur_pos += 1.0
        done = self.cur_pos >= self.end_pos
        return [self.cur_pos], 1 if done else 0, done, {}

ray.init()

ray_config = {
    "use_exec_api": True,

# === Settings for Rollout Worker processes ===
    "num_workers": 3,
    "num_envs_per_worker": 1,
    # Update the replay buffer with this many samples at once. Note that
    # this setting applies per-worker if num_workers > 1.
    "rollout_fragment_length": 128,
    "batch_mode": "complete_episodes",#"complete_episodes",

    # === Settings for the Trainer process ===
    "num_gpus": 0,
    "train_batch_size": 512,
    "model": {
        "fcnet_activation" : "relu",
        "fcnet_hiddens" : [128, 128],
    },
    # Arguments to pass to the policy optimizer. These vary by optimizer.
    "optimizer": {},

    # === Debug Settings ===
    "monitor": False,
    "log_level": "DEBUG",
    "ignore_worker_failures": False,
    "log_sys_usage": True,

    "use_pytorch": False,
    "min_iter_time_s": 1,

    "num_cpus_per_worker": 1,
    "num_gpus_per_worker": 0,
    "custom_resources_per_worker": {},
    "num_cpus_for_driver": 1,
    "memory": 0,
    "object_store_memory": 0,
    "memory_per_worker": 0,
    "object_store_memory_per_worker": 0,

 # ALGORITHM SPECIFIC
    "double_q": True,
    "n_step": 3,
    "target_network_update_freq": 1000,
    "buffer_size": 50000,
    "learning_starts": 1000,
    # Minimum env steps to optimize for per train call. This value does
    # not affect learning, only the length of iterations.
    "timesteps_per_iteration": 1000,
    "exploration_config" : {"type": "PerWorkerEpsilonGreedy"},
    "worker_side_prioritization": True,
}

env_config = {
    "env_config": {
            'corridor_length': 5,
        }
}

trainer = agents.dqn.ApexTrainer(
    env=SimpleCorridor,
    config={**env_config, **ray_config}
)

while True:
    results = trainer.train()
    print(f"Episodes Total: {results['episodes_total']}")
    if results['episodes_total'] >= 100:
        break

print(results['episodes_total'])
```

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by ericl** [comment]
Can you try with latest? It's now on by default and should be stable since it is passing the cluster release regression tests.

**Comment by ElektroChan89** [comment]
I tried it out with 0.8.6 version and there is still the same problem.

**Comment by ericl** [comment]
Oh I think I know what's going on. The `rewards` for Ape-X are only reported for the workers from the lowest third of epsilons (otherwise, we would be reporting rewards from workers with epsilon as high as 0.4, which usually are completely terrible). This affects all the related metrics such as episodes_total.

tldr: just a reporting artifact. You can however rely on timesteps_total to be accurate as those are collected independently.

**Comment by ElektroChan89** [comment]
Thanks for the answer.
An entry in the docs would be helpful.
Is this also true for other algorithms, like DQN or only for APEX?
